middleware/websocket/websocket.go

```diff
@@ -0,0 +1,220 @@
+	remoteHost, remotePort, err := net.SplitHostPort(r.RemoteAddr)
+	if err != nil {
+		return
+	}
+
+	serverHost, serverPort, err := net.SplitHostPort(r.Host)
+	if err != nil {
+		return
+	}
+
+	metavars = []string{
+		`AUTH_TYPE=`,      // Not used
+		`CONTENT_LENGTH=`, // Not used
+		`CONTENT_TYPE=`,   // Not used
+		`GATEWAY_INTERFACE=` + GatewayInterface,
+		`PATH_INFO=`,       // TODO
+		`PATH_TRANSLATED=`, // TODO
+		`QUERY_STRING=` + r.URL.RawQuery,
+		`REMOTE_ADDR=` + remoteHost,
+		`REMOTE_HOST=` + remoteHost, // Host lookups are slow - don't do them
+		`REMOTE_IDENT=`,             // Not used
+		`REMOTE_PORT=` + remotePort,
+		`REMOTE_USER=`, // Not used,
+		`REQUEST_METHOD=` + r.Method,
+		`REQUEST_URI=` + r.RequestURI,
+		`SCRIPT_NAME=` + cmdPath, // path of the program being executed
+		`SERVER_NAME=` + serverHost,
+		`SERVER_PORT=` + serverPort,
+		`SERVER_PROTOCOL=` + r.Proto,
+		`SERVER_SOFTWARE=` + ServerSoftware,
+	}
+
+	// Add each HTTP header to the environment as well
+	for header, values := range r.Header {
+		value := strings.Join(values, ", ")
+		header = strings.ToUpper(header)
+		header = strings.Replace(header, "-", "_", -1)
+		value = strings.Replace(value, "\n", " ", -1)
+		metavars = append(metavars, "HTTP_"+header+"="+value)
+	}
+
+	return
+}
+
+// reader is the guts of this package. It takes the stdin and stdout pipes
+// of the cmd we created in ServeWS and pipes them between the client and server
+// over websockets.
+func reader(conn *websocket.Conn, stdout io.ReadCloser, stdin io.WriteCloser) {
+	// Setup our connection's websocket ping/pong handlers from our const values.
+	conn.SetReadLimit(maxMessageSize)
+	conn.SetReadDeadline(time.Now().Add(pongWait))
+	conn.SetPongHandler(func(string) error { conn.SetReadDeadline(time.Now().Add(pongWait)); return nil })
+	go ticker(conn)
+
+	for {
+		msgType, r, err := conn.NextReader()
+		if err != nil {
+			if msgType == -1 {
+				return // we are done, as we got a close method.
+			}
+			panic(err) // TODO do something else here.
+		}
+
+		w, err := conn.NextWriter(msgType)
+		if err != nil {
+			panic(err) // TODO do something else here.
+		}
+
+		if _, err := io.Copy(stdin, r); err != nil {
+			panic(err) // TODO do something else here.
+		}
+
+		go func() {
+			if _, err := io.Copy(w, stdout); err != nil {
+				panic(err) // TODO do something else here.
+			}
+			if err := w.Close(); err != nil {
+				panic(err) // TODO do something else here.
+			}
+		}()
+	}
+}
+
+// ticker is start by the reader. Basically it is the method that simulates the websocket
+// between the server and client to keep it alive with ping messages.
+func ticker(conn *websocket.Conn) {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		ticker.Stop()
+		conn.WriteMessage(websocket.CloseMessage, nil)
+	}()
+
+	for { // blocking loop with select to wait for stimulation.
+		select {
+		case <-ticker.C:
+			conn.WriteMessage(websocket.PingMessage, nil)
+		}
```

Comment:
There's something missing here to check that the connection has been closed and the goroutine should return. Otherwise this will leak goroutines.